File: scripts/generate-keys.js
#!/usr/bin/env node

import crypto from 'crypto';
import fs from 'fs';
import path from 'path';

const ENV_PATH = '.env';

/**
 * Generate a secure random key of specified byte length.
 * Returns hex-encoded string (suitable for AES-256).
 */
function generateSecureKey(length = 32) {
  return crypto.randomBytes(length).toString('hex');
}

/**
 * Generate a secure JWT secret.
 * Returns base64-encoded string.
 */
function generateJWTSecret(length = 64) {
  return crypto.randomBytes(length).toString('base64url');
}

/**
 * Read existing .env file and parse key-value pairs
 */
function readEnvFile(envPath) {
  if (!fs.existsSync(envPath)) {
    return {};
  }
  
  const content = fs.readFileSync(envPath, 'utf8');
  const env = {};
  
  content.split('\n').forEach(line => {
    const trimmed = line.trim();
    if (trimmed && !trimmed.startsWith('#')) {
      const [key, ...valueParts] = trimmed.split('=');
      if (key && valueParts.length > 0) {
        env[key.trim()] = valueParts.join('=').trim().replace(/^["']|["']$/g, '');
      }
    }
  });
  
  return env;
}

/**
 * Write keys to .env file, preserving existing variables
 */
function writeKeysToEnv(aesKey, jwtSecret, envPath = ENV_PATH) {
  // Read existing .env content
  const existingEnv = readEnvFile(envPath);
  
  // Update with new keys
  existingEnv.ENCRYPTION_KEY = aesKey;
  existingEnv.JWT_SECRET = jwtSecret;
  
  // Convert back to .env format
  const envContent = Object.entries(existingEnv)
    .map(([key, value]) => `${key}=${value}`)
    .join('\n') + '\n';
  
  // Write to file
  fs.writeFileSync(envPath, envContent, 'utf8');
}

/**
 * Validate generated keys meet security requirements
 */
function validateKeys(aesKey, jwtSecret) {
  const issues = [];
  
  // AES key should be 64 hex characters (32 bytes)
  if (aesKey.length !== 64) {
    issues.push(`AES key length: ${aesKey.length} (expected 64)`);
  }
  
  if (!/^[a-f0-9]+$/i.test(aesKey)) {
    issues.push('AES key contains non-hex characters');
  }
  
  // JWT secret should be at least 32 characters
  if (jwtSecret.length < 32) {
    issues.push(`JWT secret too short: ${jwtSecret.length} (minimum 32)`);
  }
  
  return issues;
}

function main() {
  console.log('🔐 Generating secure encryption keys for Prayas platform...\n');
  
  // Generate keys
  const aes256Key = generateSecureKey(32); // 32 bytes = 256 bits
  const jwtSecretKey = generateJWTSecret(64); // 64 bytes for extra security
  
  // Validate keys
  const issues = validateKeys(aes256Key, jwtSecretKey);
  if (issues.length > 0) {
    console.error('❌ Key validation failed:');
    issues.forEach(issue => console.error(`  - ${issue}`));
    process.exit(1);
  }
  
  // Save to .env
  try {
    writeKeysToEnv(aes256Key, jwtSecretKey);
    console.log('✅ Keys generated and saved successfully!\n');
    
    // Display key information (for verification)
    console.log('Generated Keys:');
    console.log('==============');
    console.log(`ENCRYPTION_KEY: ${aes256Key}`);
    console.log(`JWT_SECRET: ${jwtSecretKey.substring(0, 16)}...[hidden]`);
    console.log(`\n📁 Keys saved to: ${ENV_PATH}`);
    
    // Security recommendations
    console.log('\n🛡️  Security Recommendations:');
    console.log('  - Keep your .env file secure and never commit it to version control');
    console.log('  - Use different keys for production and development environments');
    console.log('  - Rotate keys periodically for maximum security');
    console.log('  - Backup keys securely before rotation');
    
  } catch (error) {
    console.error('❌ Failed to save keys:', error.message);
    process.exit(1);
  }
}

// Run if called directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}

export { generateSecureKey, generateJWTSecret, writeKeysToEnv, validateKeys };

File: scripts/generate-keys.py
#!/usr/bin/env python3
"""
Secure Key Generator for Prayas Platform
Generates AES-256 encryption keys and JWT secrets compatible with Node.js crypto system.
"""

import os
import secrets
import base64
from typing import Tuple

ENV_PATH = ".env"

def generate_aes256_key() -> str:
    """
    Generate a secure AES-256 key (32 bytes = 256 bits).
    Returns hex-encoded string compatible with Node.js crypto.
    """
    return secrets.token_hex(32)  # 32 bytes = 64 hex characters

def generate_jwt_secret(length: int = 64) -> str:
    """
    Generate a secure JWT secret key.
    Returns base64url-encoded string for maximum compatibility.
    """
    return base64.urlsafe_b64encode(secrets.token_bytes(length)).decode().rstrip('=')

def read_env_file(env_path: str) -> dict:
    """Read existing .env file and parse key-value pairs."""
    if not os.path.exists(env_path):
        return {}
    
    env_vars = {}
    with open(env_path, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if line and not line.startswith('#'):
                if '=' in line:
                    key, value = line.split('=', 1)
                    env_vars[key.strip()] = value.strip().strip('"\'')
    
    return env_vars

def write_keys_to_env(aes_key: str, jwt_secret: str, env_path: str = ENV_PATH) -> None:
    """
    Save AES and JWT secret keys to .env file, preserving existing variables.
    Uses Node.js compatible variable names: ENCRYPTION_KEY and JWT_SECRET.
    """
    # Read existing .env content
    existing_env = read_env_file(env_path)
    
    # Update with new keys (using Node.js compatible names)
    existing_env['ENCRYPTION_KEY'] = aes_key
    existing_env['JWT_SECRET'] = jwt_secret
    
    # Write back to file
    with open(env_path, 'w', encoding='utf-8') as f:
        for key, value in existing_env.items():
            f.write(f"{key}={value}\n")

def validate_keys(aes_key: str, jwt_secret: str) -> list:
    """Validate generated keys meet security requirements."""
    issues = []
    
    # AES key should be 64 hex characters (32 bytes)
    if len(aes_key) != 64:
        issues.append(f"AES key length: {len(aes_key)} (expected 64)")
    
    if not all(c in '0123456789abcdefABCDEF' for c in aes_key):
        issues.append("AES key contains non-hex characters")
    
    # JWT secret should be at least 32 characters
    if len(jwt_secret) < 32:
        issues.append(f"JWT secret too short: {len(jwt_secret)} (minimum 32)")
    
    return issues

def generate_keys() -> Tuple[str, str]:
    """Generate both AES-256 and JWT keys."""
    aes_key = generate_aes256_key()
    jwt_secret = generate_jwt_secret(64)  # 64 bytes for extra security
    return aes_key, jwt_secret

def main():
    """Main function to generate and save encryption keys."""
    print("🔐 Generating secure encryption keys for Prayas platform...\n")
    
    # Generate keys
    aes256_key, jwt_secret_key = generate_keys()
    
    # Validate keys
    issues = validate_keys(aes256_key, jwt_secret_key)
    if issues:
        print("❌ Key validation failed:")
        for issue in issues:
            print(f"  - {issue}")
        return False
    
    # Save to .env
    try:
        write_keys_to_env(aes256_key, jwt_secret_key)
        print("✅ Keys generated and saved successfully!\n")
        
        # Display key information (for verification)
        print("Generated Keys:")
        print("==============")
        print(f"ENCRYPTION_KEY: {aes256_key}")
        print(f"JWT_SECRET: {jwt_secret_key[:16]}...[hidden]")
        print(f"\n📁 Keys saved to: {ENV_PATH}")
        
        # Security recommendations
        print("\n🛡️  Security Recommendations:")
        print("  - Keep your .env file secure and never commit it to version control")
        print("  - Use different keys for production and development environments") 
        print("  - Rotate keys periodically for maximum security")
        print("  - Backup keys securely before rotation")
        
        return True
        
    except Exception as error:
        print(f"❌ Failed to save keys: {error}")
        return False

if __name__ == "__main__":
    success = main()
    exit(0 if success else 1)